### PR TITLE
Unit Tests and Code Coverage

### PR DESCRIPTION
**Pull request by seanoshea**
Build available at https://circleci.com/gh/seanoshea/material-motion-runtime-android/13. Associated code coverage report available at https://codecov.io/gh/seanoshea/material-motion-runtime-android/commit/549ba9f04036f7ae4f5ad6af1911cd8aed12aa4d.

Forced a unit test failure locally and received:

```
com.google.android.material.motion.runtime.ApplicationTest > testOne[Nexus_6_API_23(AVD) - 6.0] FAILED 
> Buildijunit.framework.AssertionFailedError: expected:<1> but was:<0>
    at com.google.android.material.motion.runtime.ApplicationTest.testOne(ApplicationTest.java:32)
:library:connectedDebugAndroidTest FAILED

FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':library:connectedDebugAndroidTest'.
> There were failing tests. See the report at: file:///Users/sean.oshea/github/material-motion-runtime-android/library/build/reports/androidTests/connected/index.html

* Try:
Run with --stacktrace option to get the stack trace. Run with --info or --debug option to get more log output.

BUILD FAILED

Total time: 11.74 secs
soshea-mbpr:material-motion-runtime-android sean.oshea$ ./gradlew library:connectedAndroidTest
```

While I started off trying to get this working with `travis-ci`, I couldn't successfully get the app to build consistently as documented in https://github.com/material-motion/material-motion-runtime-android/issues/4. Switched over to `circleci` and was able to get the app to build, run the unit tests and produce code coverage. Not sure if the original intent of the Github issue was mostly to get continuous integration working or if you're ok with a `circleci` continuous integration implementation.

cc @pingpongboss on this as we had a chat about this on discord.

### CODE REVIEW

File: sample/build.gradle [hunk 1]
```diff
@@ -17,6 +17,9 @@ android {
   }
 
   buildTypes {
+    debug {
```
Comment (seanoshea) [inline sample/build.gradle]: Hmmm. Not 100% sure I need this here. Will try a fresh build without this config option. I'm assuming we really only care about the coverage for the actual library vs the sample application?

File: README.md [hunk 2]
```diff
@@ -1,4 +1,6 @@
 # Material Motion Android Runtime
+[![Build Status](https://circleci.com/gh/material-motion/material-motion-runtime-android.svg?style=shield)](https://circleci.com/gh/material-motion/material-motion-runtime-android)
```
Comment (pingpongboss) [inline README.md]: how do we unbreak these svg and links

**Comment by pingpongboss** [comment]
This is great! Just to clarify, the original intention of the issue is to research how to enable CI for our repos. TravisCI was named as a potential solution, and it is slightly preferred over other CI solutions because our objc/swift counterparts use it. But it looks like enabling CircleCI is a great solution.

File: sample/build.gradle (hunk 1 above)
Comment (pingpongboss) [inline sample/build.gradle]: Right - we're only interested in testing the actual library.

File: README.md (hunk 2 above)
Comment (seanoshea) [inline README.md]: @pingpongboss - someone on the `material-motion` team would need to log into `circleci` and grant it access to build the repo. Once the PR is merged to develop, those badges should work ok. https://github.com/seanoshea/BluePlaquesLondon-Android/blob/develop/README.md is one example of how this works for other projects.

**Comment by seanoshea** [comment]
@pingpongboss - yep, its unfortunate that I couldn't get travis to work as you mentioned other `material-motion` counterparts use travis. According to https://docs.travis-ci.com/user/languages/android, their Android support is still in beta though (and has been for the past year or so) - perhaps its not a priority at this time. circleci seems to work fine though.

**Comment by seanoshea** [comment]
Please hold on this PR. There are some reservations in Discord about granting circleci read/write access to repos. Will try to resurrect travis-ci efforts.

**Comment by seanoshea** [comment]
Went back to the drawing board on this. I think I have the last three travis builds passing:

```
https://travis-ci.org/seanoshea/material-motion-runtime-android/builds/152863695
https://travis-ci.org/seanoshea/material-motion-runtime-android/builds/152862160
https://travis-ci.org/seanoshea/material-motion-runtime-android/builds/152855028
```

cc @pingpongboss @jverkoey

**Comment by jverkoey** [comment]
Lookin' stable!

**Comment by seanoshea** [comment]
@jverkoey The code in `travis.yml` is exactly the same as it was prior to me previously giving up on travis and trying out circleci. So, the fact that it's stable doesn't have anything to do with changes as part of this PR. Either the build is happening on a more favorable machine, or travis have fixed something with their Android implementation.

But, yeah - it's looking a little more stable at this time.

**Comment by jverkoey** [comment]
Oh, that's worrying then hah. Is there a specific question we could ask the Travis CI team to get more information about why it might be flaking/stable? They're decently responsive on twitter.

**Comment by seanoshea** [comment]
Tried pinging them with https://twitter.com/seanoshea/status/763155140404731904 and https://twitter.com/seanoshea/status/763155327101575168, but no response.

https://github.com/material-motion/material-motion-runtime-android/blob/ef1052b5308eb8e860e9eedfb72a773e923a581f/.travis.yml is where I disabled travis-ci the first time. https://raw.githubusercontent.com/seanoshea/material-motion-runtime-android/4f0f228bc6b0af8e3543a9c1ffafe4371b3fc552/.travis.yml is what it looks like now & you'll notice that they are exactly the same.

I think we have a couple of options here:
1. Reach out to travis and see whether they have any information on whether any changed on their end to help support Android development a bit more. Perhaps ask them if they have any plans to make Android a 1st class citizen on their platform for opensource projects.
2. Convert the tests to Robolectric which doesn't require a simulator and might help the tests run a little more consistently & continue to use travis.
3. Go back and use circleci as that appeared to be always consistent.

**Comment by jverkoey** [comment]
As a point of reference, here's facebook/rebound's .travis.yml file: https://raw.githubusercontent.com/facebook/rebound/master/.travis.yml

Is there anything going on in there that's different to how we're approaching this?

**Comment by seanoshea** [comment]
From their latest build, it appears that they are just invoking `smokeTest` which looks like https://github.com/facebook/rebound/blob/master/build.gradle#L71. I believe this just builds a jar and doesn't require them to fire up an emulator to run unit tests.

Actually, looking at https://s3.amazonaws.com/archive.travis-ci.org/jobs/152636483/log.txt, I don't believe they are running tests as part of their CI flow at all. Hmmmm.

Had a look at a few Square libraries (I really like their opensource Android stuff) and it looks like they generally rely on travis to build their default gradle target & don't actually ever fire up the emulator:

```
https://github.com/square/spoon
https://github.com/square/picasso
https://github.com/square/dagger
https://github.com/square/leakcanary
https://github.com/square/android-times-square
```

**Comment by seanoshea** [comment]
Unsure of how best to proceed with this PR. In https://github.com/material-motion/material-motion-runtime-android/issues/4, it was suggested that we could possibly upgrade gradle to help with the stability. Happy to give that a try if you think it would help this PR.

The other option on the table would be to transition over to robolectric tests which doesn't require the emulator to be running to actually run the tests.

So, the tradeoff is between running standard Android tests & potentially running into issues with getting the emulator up and running in the future vs using a non-standard Android unit testing framework in a mocked up emulator.

**Comment by jverkoey** [comment]
I'm inclined for us to move forward with this PR and to resolve flakiness/issues over time as they appear. @pingpongboss thoughts?

**Comment by pingpongboss** [comment]
Right, I think this PR is a step in the right direction. Let's merge it. @seanoshea is this ready to be merged?

**Comment by seanoshea** [comment]
@pingpongboss @jverkoey - Yes, I think this is good to go. Gave it one final go with https://travis-ci.org/seanoshea/material-motion-runtime-android/builds/154627845 and it passed, so I think any flakiness has disappeared for the time being.

Once this is merged in, I'll check and see if I can get travis and codecov to pick up the builds.

**Pull request merged**